xdsl/parser.py

```diff
@@ -1,1335 +1,1924 @@
+    comma = re.compile(",")
+
+
+class BaseParser(ABC):
+    """
+    Basic recursive descent parser.
 
-        return res
+    methods marked try_... will attempt to parse, and return None if they failed. If they return None
+    they must make sure to restore all state.
 
-    def parse_optional_dict_entry(
-        self,
-        parse_optional_key: Callable[[], K | None],
-        parse_optional_value: Callable[[], V | None],
-    ) -> tuple[K, V] | None:
-        if (key := parse_optional_key()) is None:
-            return None
-        self.parse_char("=")
-        if (value := parse_optional_value()) is None:
-            raise ParserError(self._pos,
-                              'Expected dictionary value after `key=`')
+    methods marked must_... will do greedy parsing, meaning they consume as much as they can. They will
+    also throw an error if the think they should still be parsing. e.g. when parsing a list of numbers
+    separated by '::', the following input will trigger an exception:
+        1::2::
+    Due to the '::' present after the last element. This is useful for parsing lists, as a trailing
+    separator is usually considered a syntax error there.
 
-        return key, value
+    You can turn a try_ into a must_ by using expect(try_parse_..., error_msg)
 
-    def parse_optional_block_argument(
-            self,
-            skip_white_space: bool = True) -> tuple[str, Attribute] | None:
-        name = self.parse_optional_ssa_name(skip_white_space=skip_white_space)
-        if name is None:
-            return None
-        self.parse_char(":")
-        typ = self.parse_attribute()
-        # TODO how to get the id?
-        return name, typ
-
-    def parse_optional_named_block(self,
-                                   skip_white_space: bool = True
-                                   ) -> Block | None:
-        if self.parse_optional_char("^",
-                                    skip_white_space=skip_white_space) is None:
-            return None
-        block_name = self.parse_alpha_num(skip_white_space=False)
-        if block_name in self._blocks:
-            block = self._blocks[block_name]
+    You can turn a must_ into a try_ by wrapping it in tokenizer.backtracking()
+
+    must_ type parsers are preferred because they are explicit about their failure modes.
+    """
+
+    ctx: MLContext
+    """xDSL context."""
+
+    ssaValues: dict[str, SSAValue]
+    blocks: dict[str, Block]
+    forward_block_references: dict[str, list[Span]]
+    """
+    Blocks we encountered references to before the definition (must be empty after parsing of region completes)
+    """
+
+    T_ = TypeVar("T_")
+    """
+    Type var used for handling function that return single or multiple Spans. Basically the output type
+    of all try_parse functions is T_ | None
+    """
+
+    allow_unregistered_ops: bool
+
+    def __init__(self,
+                 ctx: MLContext,
+                 input: str,
```

Comment:
This has the implicit assumption that each Input has to be parsed with its own Parser, I would prefer having one Parser that can handle a lot of inputs, but I don't know the structure well enough to understand whether that is even doable. Maybe @AntonLydike could shed some light into this?

Reply:
Wait, I'm having difficulty understanding your question. Are you saying you want to be able to parse multiple files *of the same syntax* with one instance of *Parser*?

Or do you want to have one parser class that is able to parse xdsl and mlir code?

Reply:
Yes, same syntax, same parser, but different strings.

Reply:
I think it would be cleaner to instantiate a new parser for each input, rather than making sure we change all the state all the time we load new input. I think this is not a worthwhile optimization.

I'm interested, why would you want to have one instance parse multiple inputs?

Reply:
I just feel like it intuitively makes sense to me, but if that is too much headache for book keeping, I am fine with it.

I must say though i don't quite understand the API then. Why split construction from running, if there is only one result this can give?

Reply:
That's just the interface of the old parser, it's also instantiated with an input, but parse is called separately. I feel that that is much more natural than running the parser on instantiation... If you don't like the interface you can always wrap it in a function like

```python
def parse(input):
    parser = Parser(input)
    return parser.parse_module()
```

or something like that.

Reply:
Sure, let's leave it as is for the moment :+1: